16: sitemap index skeleton

kirstyburgoine opened the pull request:
### Issue Number
https://github.com/swissspidy/wp-sitemaps/issues/16

### Description
Basic skeleton for sitemap_index.xml

### Type of change
Please select the relevant options:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Steps to test
- Activate the plugin
- View yourdomain/sitemap_index.xml

### Acceptance criteria
- [x] My code follows WordPress coding standards.
- [x] I have performed a self-review of my own code.
- [ ] If the changes are visual, I have cross browser / device tested.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [x] I have added test instructions that prove my fix is effective or that my feature works.

kirstyburgoine committed:
16: sitemap index skeleton

kirstyburgoine committed:
16: Remove additional space

svandragt commented on core-sitemaps.php:
```diff
@@ -18,3 +18,7 @@
 
 // Your code starts here.
+
+require_once dirname( __FILE__ ) . '/sitemaps-index.php';
```
This is the same as the code below but simpler.
```suggestion
require_once __DIR__ . '/sitemaps-index.php';
```

kirstyburgoine committed:
16: Lint

svandragt commented on sitemaps-index.php:
```diff
@@ -0,0 +1,78 @@
+		global $wp;
+		$wp->add_query_var( 'sitemap' );
+
+		add_rewrite_rule( 'sitemap_index\.xml$', 'index.php?sitemap=sitemap', 'top' );
```
Best prepend the regex with ^ as this will avoid accendental matching on paths ending in the xml file. Sitemap convention dictates the main sitemap is found at sitemap.xml, what's the thinking behind _index?
```suggestion
		add_rewrite_rule( '^sitemap\.xml$', 'index.php?sitemap=sitemap', 'top' );
```

svandragt commented on sitemaps-index.php:
```diff
@@ -0,0 +1,78 @@
+	 */
+	public function url_rewrites() {
+		global $wp;
+		$wp->add_query_var( 'sitemap' );
```
Instead of using the global in this way, the following code can be used that has more safety checks around it.
`add_rewrite_tag('%sitemap%','sitemap');`

svandragt commented on sitemaps-index.php:
```diff
@@ -0,0 +1,78 @@
+
+		if ( ! empty( $sitemap_index ) ) {
+
+			header( 'Content-type: text/xml; charset=' );
```
Based on the output below the sitemap contains UTF-8. In that case application/xml is the better content-type. see https://stackoverflow.com/questions/3272534/what-content-type-value-should-i-send-for-my-xml-sitemap, and I think this line misses the charset value?
```suggestion
			header( 'Content-type: application/xml; charset=UTF-8' );
```

svandragt reviewed (changes requested):
Good code, very clear. I've added feedback based on other sitemap implementations for your consideration. I'll definiately want to adjust the PR for Posts Sitemaps to work in an object orientated manner.

As an aside, we'd want to flush the rewrite rules when they have changed. In my experience this is best done hooking into init, if update_option('WP_SITEMAPS_REWRITE_VERSION') changed, as this avoids unnecessary flushes and race conditions with checking option values.

kirstyburgoine committed:
Use __DIR__ for require_once

Co-Authored-By: Sander van Dragt <sander@humanmade.com>

kirstyburgoine committed:
prepend regex with ^ for rewrite rule

Co-Authored-By: Sander van Dragt <sander@humanmade.com>

kirstyburgoine committed:
Update content-type and charset

Co-Authored-By: Sander van Dragt <sander@humanmade.com>

kirstyburgoine committed:
16: Remove global $wp

kirstyburgoine committed:
16: Move sitemaps index into /inc/ folder

kirstyburgoine committed:
16: Lint

joemcgill commented on inc/sitemaps-index.php:
```diff
@@ -0,0 +1,73 @@
+	 * Class constructor.
+	 */
+	public function __construct() {
+		add_action( 'init', array( $this, 'url_rewrites' ), 99 );
```
Let's move actions/filters out of the constructor and into a separate `bootstrap()` method. Even though you will find this pattern common in WordPress code, I generally consider it an anti-pattern that should be avoided so that you can instantiate the class without introducing side effects. It also makes your code more testable because you can load your class without hooking into all the WordPress functionality.

joemcgill reviewed (changes requested):
This is looking like a good base to start from. Let's get this in and iterate from here. I only have one suggestion about where to hook actions/filters, otherwise, It think it's good to go.

joemcgill commented on core-sitemaps.php:
```diff
@@ -18,3 +18,7 @@
 
 // Your code starts here.
+
+require_once __DIR__ . '/inc/sitemaps-index.php';
```
Let's rename this file to class-sitemaps-index.php, per https://engineering.hmn.md/standards/style/php/#file-naming.

kirstyburgoine committed:
16: Rename to class-sitemaps-index.php

swissspidy commented on inc/sitemaps-index.php:
```diff
@@ -0,0 +1,73 @@
+		$sitemap_index = get_query_var( 'sitemap' );
+
+		if ( ! empty( $sitemap_index ) ) {
+			header( 'Content-type: application/xml; charset=UTF-8' );
```
Should this use `get_option( 'blog_charset' )`?

joemcgill replied:
No, based on the [sitempas.org spec](https://www.sitemaps.org/protocol.html):

> The Sitemap protocol format consists of XML tags. All data values in a Sitemap must be entity-escaped. The file itself must be UTF-8 encoded.

swissspidy replied:
Great!

kirstyburgoine commented:
@googlebot I consent

kirstyburgoine committed:
16: Move filters to bootstrap()

kirstyburgoine committed:
16: lint

kirstyburgoine committed:
16: Fix white screen of doom

kirstyburgoine committed:
16: Update function comments

kirstyburgoine committed:
16 Update output_sitemap() to redirect to template

kirstyburgoine committed:
Merge branch 'master' into feature/16-index-sitemap

joemcgill commented on inc/class-sitemaps-index.php:
```diff
@@ -0,0 +1,73 @@
+	public function output_sitemap( $template ) {
+		$sitemap_index = get_query_var( 'sitemap' );
+
+		if ( ! empty( $sitemap_index ) ) {
```
Let's print the XML directly here rather than doing a redirect. We'll end up splitting this out anyway, but we should be able to do so without a 301.

joemcgill commented on inc/class-sitemaps-index.php:
```diff
@@ -0,0 +1,73 @@
+	 * @param string  $template The template to return. Either custom XML or default.
+	 * @return string
+	 *
+	 * @todo Review how the sitemap files are built and placed in the root of the site.
```
Do we need these `todos` anymore? If so, let's open new tickets and remove the comments from the code.

joemcgill commented on inc/class-sitemaps-index.php:
```diff
@@ -0,0 +1,73 @@
+	 */
+	public function url_rewrites() {
+		add_rewrite_tag( '%sitemap%','sitemap' );
+		add_rewrite_rule( 'sitemap_index\.xml$', 'index.php?sitemap=sitemap', 'top' );
```
Let's change this to `sitemap=index` or `sitemap=sitemap_index` so it's more descriptive when we check for this value later in our code.

joemcgill reviewed (changes requested):
Hi @kirstyburgoine, I've got a few last changes and I think this is ready to go.

kirstyburgoine committed:
16: Final feedback bits

- Removed @todo’s
- Update sitemap=sitemap to sitemap=sitemap_index
- Print xml directly instead of redirecting (this reverts back to earlier commits for potential iteration)

swissspidy commented on inc/sitemap_index.xml:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This feel doesn't seem to be needed?

kirstyburgoine committed:
16: Delete xml file

kirstyburgoine committed:
16: Use template_redirect instead template_include

kirstyburgoine committed:
16: Remove @todo

kirstyburgoine committed:
16: Remove $sitemap_content and lint

kirstyburgoine committed:
16: Update docblock for output_sitemap()

joemcgill reviewed (approved):
Thanks @kirstyburgoine. I like the simplification for now. Looks like it's working.

svandragt reviewed (approved):
Great, please merge this!
